src/datatypes/sipath.c

```diff
@@ -0,0 +1,173 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "sipath.h"
+#include "../util/rmalloc.h"
+#include "../util/arr.h"
+#include "array.h"
+
+SIValue SIPath_New(Path *p) {
+	SIValue path;
+	path.ptrval = rm_malloc(sizeof(Path));
+	*((Path *)path.ptrval) = Path_Clone(*p);
+	path.type = T_PATH;
+	path.allocation = M_SELF;
+	return path;
+}
+
+SIValue SIPath_Clone(SIValue p) {
+	return SIPath_New((Path *)p.ptrval);
+}
+
+SIValue SIPath_ToList(SIValue p) {
+	size_t nodeCount = SIPath_NodeCount(p);
+	size_t edgeCount = SIPath_Length(p);
+	SIValue array = SI_Array(nodeCount + edgeCount);
+	for(size_t i = 0; i < nodeCount - 1 ; i++) {
+		SIValue node = SIPath_GetNode(p, i);
+		SIArray_Append(&array, node);
+		SIValue edge = SIPath_GetRelationship(p, i);
+		SIArray_Append(&array, edge);
+	}
+	if(nodeCount > 0) {
+		SIValue node = SIPath_GetNode(p, nodeCount - 1);
+		SIArray_Append(&array, node);
+	}
+	return array;
+}
+
+SIValue SIPath_Relationships(SIValue p) {
+	Path path = *((Path *) p.ptrval);
+	uint edgeCount = Path_EdgeCount(path);
+	SIValue array = SIArray_New(edgeCount);
+	for(uint i = 0; i < edgeCount; i++) {
+		SIArray_Append(&array, SI_Edge(&path.edges[i]));
+	}
+	return array;
+}
+
+SIValue SIPath_GetRelationship(SIValue p, size_t i) {
+	assert(i < SIPath_Length(p) && i >= 0);
+	Path path = *((Path *) p.ptrval);
+	return SI_Edge(&path.edges[i]);
+}
+
+SIValue SIPath_Nodes(SIValue p) {
+	Path path = *((Path *) p.ptrval);
+	uint nodeCount = Path_NodeCount(path);
+	SIValue array = SIArray_New(nodeCount);
+	for(uint i = 0; i < nodeCount; i++) {
+		SIArray_Append(&array, SI_Node(&path.nodes[i]));
+	}
+	return array;
+}
+
+SIValue SIPath_GetNode(SIValue p, size_t i) {
+	assert(i < SIPath_NodeCount(p) && i >= 0);
+	Path path = *((Path *) p.ptrval);
+	return SI_Node(&path.nodes[i]);
+}
+
+size_t SIPath_Length(SIValue p) {
+	Path path = *((Path *) p.ptrval);
+	return Path_Len(path);
+}
+
+size_t SIPath_NodeCount(SIValue p) {
+	Path path = *((Path *) p.ptrval);
+	return Path_NodeCount(path);
+}
+
+size_t SIPath_EdgeCount(SIValue p) {
+	Path path = *((Path *) p.ptrval);
+	return Path_EdgeCount(path);
+}
+
+XXH64_hash_t SIPath_HashCode(SIValue p) {
```

Comment:
This is not called currently, should add a case at value.c:571

Reply:
done